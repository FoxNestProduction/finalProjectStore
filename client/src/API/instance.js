import axios from 'axios';

let store;

export const injectStore = (_store) => {
  store = _store;
};

export const instance = axios.create({
  baseURL: process.env.REACT_APP_API_URL,
});

instance.interceptors.request.use((config) => {
  if (store.getState().authorization && store.getState().authorization.token) {
    // eslint-disable-next-line no-param-reassign
    config.headers.authorization = store.getState().authorization.token;
  }
  return config;
}, (error) => {
  return Promise.reject(error);
});
