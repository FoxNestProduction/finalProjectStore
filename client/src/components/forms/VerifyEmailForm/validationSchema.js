import { object, string } from 'yup';

const validationSchema = object({
  email: string()
    .required('Email is required')
    .email('Invalid email format'),
});

export default validationSchema;
