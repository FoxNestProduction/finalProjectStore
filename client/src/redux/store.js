import { configureStore, getDefaultMiddleware, combineReducers } from '@reduxjs/toolkit';
// todo: правила для redux-persist
// eslint-disable-next-line import/no-extraneous-dependencies
import storage from 'redux-persist/lib/storage';
// eslint-disable-next-line import/no-extraneous-dependencies
import { persistReducer } from 'redux-persist';
import modalSlice from './slices/modalSlice';
import productsSlice from './slices/productsSlice';
import authorizationSlice from './slices/authorizationSlice';
import userSlice from './slices/userSlice';
import partnersSlice from './slices/partnersSlice';
import searchSlice from './slices/searchSlice';
import errorSlice from './slices/errorSlice';
import reviewsSlice from './slices/reviewsSlice';
import favouriteSlice from './slices/favouriteSlice';

const authPersistConfig = {
  key: 'authorization',
  version: 1,
  storage,
};

const userPersistConfig = {
  key: 'user',
  version: 1,
  storage,
};

const searchPersistConfig = {
  key: 'search',
  version: 1,
  storage,
};
const favouritePersistConfig = {
  key: 'favorites',
  version: 1,
  storage,
};

/*
    todo: - ЗВЕРНІТЬ УВАГУ!
      якщо вам потрібно щоб ваш стейт не зникав після перезавантаження
      то юзайте підхід як authorization і user,
      якщо ж ні то варіант написання як для modal всередині reducer
      Якщо що пишіть @Ihor_Kacher
*/
const reducer = combineReducers({
  error: errorSlice,
  modal: modalSlice,
  partners: partnersSlice,
  authorization: persistReducer(authPersistConfig, authorizationSlice),
  user: persistReducer(userPersistConfig, userSlice),
  products: productsSlice,
  search: persistReducer(searchPersistConfig, searchSlice),
  reviews: reviewsSlice,
  favourites: persistReducer(favouritePersistConfig, favouriteSlice),
});

const middleware = getDefaultMiddleware({
  serializableCheck: false,
});

const store = configureStore({
  reducer,
  middleware,
});

export default store;
