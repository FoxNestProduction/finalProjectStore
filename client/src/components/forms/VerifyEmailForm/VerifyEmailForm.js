import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Formik, Form } from 'formik';
import Box from '@mui/material/Box';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import EmailIcon from '@mui/icons-material/Email';
import validationSchema from './validationSchema';
import Input from '../../inputs/Input/Input';
import { instance } from '../../../API/instance';
import { closeModal, setContent } from '../../../redux/slices/modalSlice';
import { setIsSendMail } from '../../../redux/slices/authorizationSlice';
import { flexcenter, mainTitle, legend, inputsWrapper, signInBtn } from './styles';
import SuccessfulLetter from '../../SuccessfulLetter/SuccessfulLetter';

const VerifyEmailForm = () => {
  const dispatch = useDispatch();

  const initialValues = {
    email: '',
  };

  const [loading, setLoading] = useState(false);
  const [error, setError] = useState('');

  const handleSubmit = async (email) => {
    try {
      setLoading(true);
      setError('');
      const response = await instance.post('/customers/forgot-password', email);
      if (response.status === 200) {
        dispatch(setIsSendMail(true));
        dispatch(setContent(<SuccessfulLetter />));
      }
    } catch (err) {
      console.log('Error sending mail: ', err);
      setError(err.response.data.message);
    } finally {
      setLoading(false);
    }
  };

  return (
    <Box
      component="section"
      sx={{
        ...flexcenter,
        width: '100%',
        bgcolor: 'common.white',
      }}
    >
      <Typography
        variant="h2"
        component="h1"
        sx={mainTitle}
      >
        Forgot Password
      </Typography>
      <Typography
        variant="body1"
        sx={legend}
      >
        Enter your e-mail to reset password
      </Typography>
      <Formik
        initialValues={initialValues}
        onSubmit={handleSubmit}
        validationSchema={validationSchema}
      >
        {({ isValid }) => (
          <Form>
            <Box
              sx={flexcenter}
            >
              <Box
                sx={{
                  ...flexcenter,
                  ...inputsWrapper,
                }}
              >
                <Input
                  error={error}
                  type="email"
                  name="email"
                  id="loginEmail"
                  placeholder="Enter your e-mail"
                  label="E-mail"
                  icon={<EmailIcon />}
                />
              </Box>
              <Button
                variant="contained"
                sx={signInBtn}
                type="submit"
                disabled={!isValid || loading}
              >
                Continue
              </Button>
            </Box>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default VerifyEmailForm;
