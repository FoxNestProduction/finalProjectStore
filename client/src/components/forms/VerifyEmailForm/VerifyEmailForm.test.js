/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/no-await-sync-query */
import React from 'react';
import { Provider, useDispatch } from 'react-redux';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import VerifyEmailForm from './VerifyEmailForm';
import store from '../../../redux/store';
import { instance } from '../../../API/instance';

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(),
}));

jest.mock('../../../API/instance', () => ({
  instance: {
    post: jest.fn(),
  },
}));

const mockDispatch = jest.spyOn(require('react-redux'), 'useDispatch');

describe('VerifyEmailForm Component', () => {
  mockDispatch.mockReturnValueOnce(jest.fn());
  // Testing empty email input
  test('displays error messages correctly', async () => {
    render(
      <Provider store={store}>
        <VerifyEmailForm />
      </Provider>,
    );
    const button = screen.getByText('Continue');
    fireEvent.click(button);

    await screen.findByText('Email is required');

    expect(screen.getByText('Email is required')).toBeInTheDocument();
  });

  // Testing wrong email
  test('displays error messages with wront email type', async () => {
    render(
      <Provider store={store}>
        <VerifyEmailForm />
      </Provider>,
    );
    const input = screen.getByPlaceholderText('Enter your e-mail');
    const button = screen.getByText('Continue');

    fireEvent.change(input, { target: { value: 'wrongEmail' } });
    fireEvent.click(button);

    await screen.findByText('Invalid email format');
    expect(screen.getByText('Invalid email format')).toBeInTheDocument();
  });

  test('handles form submission', async () => {
    const dispatch = jest.fn();
    mockDispatch.mockReturnValueOnce(dispatch);
    instance.post.mockResolvedValueOnce({ status: 200 });
    render(
      <Provider store={store}>
        <VerifyEmailForm />
      </Provider>,
    );

    const emailInput = screen.getByPlaceholderText('Enter your e-mail');
    const button = screen.getByText('Continue');

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.click(button);

    await waitFor(() => {
      expect(instance.post).toHaveBeenCalledTimes(1);
    });
    await waitFor(() => {
      expect(instance.post).toHaveBeenCalledWith('/customers/forgot-password', {
        email: 'test@example.com',
      });
    });
    await waitFor(() => {
      expect(useDispatch).toHaveBeenCalledTimes(2);
    });
  });

  //   Snapshot
  test('renders VerifyEmailForm correctly', async () => {
    const { asFragment } = render(
      <Provider store={store}>
        <VerifyEmailForm />
      </Provider>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
