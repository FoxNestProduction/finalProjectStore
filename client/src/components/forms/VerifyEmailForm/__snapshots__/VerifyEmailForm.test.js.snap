// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerifyEmailForm Component renders VerifyEmailForm correctly 1`] = `
<DocumentFragment>
  <section
    class="MuiBox-root css-1snrlp5"
  >
    <h1
      class="MuiTypography-root MuiTypography-h2 css-1tzbn0g-MuiTypography-root"
    >
      Forgot Password
    </h1>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1g2715c-MuiTypography-root"
    >
      Enter your e-mail to reset password
    </p>
    <form
      action="#"
    >
      <div
        class="MuiBox-root css-dvxtzn"
      >
        <div
          class="MuiBox-root css-lx50gm"
        >
          <div
            class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="loginEmail"
            >
              E-mail
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart css-1rmmbqq-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1c8d5n4-MuiInputAdornment-root"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="EmailIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4-8 5-8-5V6l8 5 8-5v2z"
                  />
                </svg>
              </div>
              <input
                aria-describedby="helper-text"
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart css-1o9s3wi-MuiInputBase-input-MuiOutlinedInput-input"
                id="loginEmail"
                name="email"
                placeholder="Enter your e-mail"
                type="email"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-14lo706"
                >
                  <span>
                    E-mail
                  </span>
                </legend>
              </fieldset>
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
              id="helper-text"
            />
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1bkqhcu-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Continue
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </form>
  </section>
</DocumentFragment>
`;
